github/github.go

```diff
@@ -332,8 +333,31 @@ func (c *Client) Rate() Rate {
 // JSON decoded and stored in the value pointed to by v, or returned as an
 // error if an API error has occurred.  If v implements the io.Writer
 // interface, the raw response body will be written to v, without attempting to
-// first decode it.
+// first decode it.  If rate limit is exceeded and reset time is in the future,
+// Do returns *RateLimitError immediately without making a network API call.
 func (c *Client) Do(req *http.Request, v interface{}) (*Response, error) {
+	rateLimitCategory := category(req.URL.Path)
```

Comment:
What do you think about pulling this entire section (~22 lines) into its own separate function call?

As it stands, I think it draws attention away from someone looking at `Do` to find out what it does... but if it were in a separate function like `checkRateLimit` (or whatever you think is better) then I think this section would be much easier to read.